File: packages/domain/src/voucher.ts
import type { CreditPackage } from './packages';
import type { TransitionMap } from './shared/machine';
import type { Result } from './shared/result';
import { err, ok } from './shared/result';
import type { IsoDate } from './shared/time';
import { isAfter, isBefore } from './shared/time';

export const VOUCHER_STATUSES = ['DRAFT', 'SCHEDULED', 'ACTIVE', 'EXPIRED', 'DISABLED'] as const;
export type VoucherStatus = (typeof VOUCHER_STATUSES)[number];

export const VOUCHER_TRANSITIONS: TransitionMap<VoucherStatus> = {
  DRAFT: ['SCHEDULED', 'ACTIVE', 'DISABLED'],
  SCHEDULED: ['ACTIVE', 'DISABLED'],
  ACTIVE: ['EXPIRED', 'DISABLED'],
  EXPIRED: [],
  DISABLED: ['ACTIVE'],
};

export interface Voucher {
  id: string;
  code: string;
  type: 'FIXED_IDR' | 'PERCENT';
  value: number;
  startsAt: IsoDate;
  endsAt: IsoDate;
  usageLimit: number | null;
  perMemberLimit: number | null;
  eligibleSegment: 'ALL' | 'NEW_MEMBERS';
  /** null = applicable to every package. */
  applicablePackageIds: string[] | null;
  status: VoucherStatus;
  createdAt: IsoDate;
}

/** Redemptions are separate records: VOUCHER → REDEMPTION → PAYMENT. */
export interface VoucherRedemption {
  id: string;
  voucherId: string;
  memberId: string;
  paymentId: string;
  discountIdr: number;
  createdAt: IsoDate;
}

export type VoucherRejection =
  | 'NOT_ACTIVE'
  | 'NOT_STARTED'
  | 'ENDED'
  | 'USAGE_LIMIT_REACHED'
  | 'PER_MEMBER_LIMIT_REACHED'
  | 'PACKAGE_NOT_ELIGIBLE'
  | 'SEGMENT_NOT_ELIGIBLE';

export function validateVoucher(args: {
  voucher: Voucher;
  pkg: CreditPackage;
  memberIsNew: boolean;
  memberRedemptionCount: number;
  totalRedemptionCount: number;
  now: IsoDate;
}): Result<{ discountIdr: number }, { reason: VoucherRejection }> {
  const { voucher, pkg, now } = args;
  if (voucher.status !== 'ACTIVE') return err({ reason: 'NOT_ACTIVE' });
  if (isBefore(now, voucher.startsAt)) return err({ reason: 'NOT_STARTED' });
  if (isAfter(now, voucher.endsAt)) return err({ reason: 'ENDED' });
  if (voucher.usageLimit !== null && args.totalRedemptionCount >= voucher.usageLimit)
    return err({ reason: 'USAGE_LIMIT_REACHED' });
  if (voucher.perMemberLimit !== null && args.memberRedemptionCount >= voucher.perMemberLimit)
    return err({ reason: 'PER_MEMBER_LIMIT_REACHED' });
  if (voucher.applicablePackageIds !== null && !voucher.applicablePackageIds.includes(pkg.id))
    return err({ reason: 'PACKAGE_NOT_ELIGIBLE' });
  if (voucher.eligibleSegment === 'NEW_MEMBERS' && !args.memberIsNew)
    return err({ reason: 'SEGMENT_NOT_ELIGIBLE' });

  const discountIdr =
    voucher.type === 'FIXED_IDR'
      ? Math.min(voucher.value, pkg.priceIdr)
      : Math.round((pkg.priceIdr * voucher.value) / 100);
  return ok({ discountIdr });
}

File: internal/domain/voucher.go
package domain

import (
	"math"
	"time"
)

// VoucherStatus is the publication state of a discount code.
type VoucherStatus string

const (
	VoucherDraft     VoucherStatus = "DRAFT"
	VoucherScheduled VoucherStatus = "SCHEDULED"
	VoucherActive    VoucherStatus = "ACTIVE"
	VoucherExpired   VoucherStatus = "EXPIRED"
	VoucherDisabled  VoucherStatus = "DISABLED"
)

// VoucherTransitions allow a disabled code to be switched back on, but an
// expired one is done: reissue instead.
var VoucherTransitions = TransitionMap[VoucherStatus]{
	VoucherDraft:     {VoucherScheduled, VoucherActive, VoucherDisabled},
	VoucherScheduled: {VoucherActive, VoucherDisabled},
	VoucherActive:    {VoucherExpired, VoucherDisabled},
	VoucherExpired:   {},
	VoucherDisabled:  {VoucherActive},
}

// VoucherType is how the discount is calculated.
type VoucherType string

const (
	VoucherFixedIDR VoucherType = "FIXED_IDR"
	VoucherPercent  VoucherType = "PERCENT"
)

// VoucherSegment limits who may redeem a code.
type VoucherSegment string

const (
	SegmentAll        VoucherSegment = "ALL"
	SegmentNewMembers VoucherSegment = "NEW_MEMBERS"
)

// Voucher is a discount code applied at checkout.
type Voucher struct {
	ID              string         `json:"id"`
	Code            string         `json:"code"`
	Type            VoucherType    `json:"type"`
	Value           int64          `json:"value"`
	StartsAt        time.Time      `json:"startsAt"`
	EndsAt          time.Time      `json:"endsAt"`
	UsageLimit      *int           `json:"usageLimit"`
	PerMemberLimit  *int           `json:"perMemberLimit"`
	EligibleSegment VoucherSegment `json:"eligibleSegment"`
	// ApplicablePackageIDs nil means the code works on every package.
	ApplicablePackageIDs []string      `json:"applicablePackageIds"`
	Status               VoucherStatus `json:"status"`
	CreatedAt            time.Time     `json:"createdAt"`
}

// VoucherRedemption records one use. Redemptions are separate rows rather than
// a counter on the voucher, so usage is auditable per member and per payment.
type VoucherRedemption struct {
	ID          string    `json:"id"`
	VoucherID   string    `json:"voucherId"`
	MemberID    string    `json:"memberId"`
	PaymentID   string    `json:"paymentId"`
	DiscountIDR int64     `json:"discountIdr"`
	CreatedAt   time.Time `json:"createdAt"`
}

// VoucherRejection is why a code was refused, surfaced to the client as
// VOUCHER_<REASON>.
type VoucherRejection string

const (
	VoucherNotActive             VoucherRejection = "NOT_ACTIVE"
	VoucherNotStarted            VoucherRejection = "NOT_STARTED"
	VoucherEnded                 VoucherRejection = "ENDED"
	VoucherUsageLimitReached     VoucherRejection = "USAGE_LIMIT_REACHED"
	VoucherPerMemberLimitReached VoucherRejection = "PER_MEMBER_LIMIT_REACHED"
	VoucherPackageNotEligible    VoucherRejection = "PACKAGE_NOT_ELIGIBLE"
	VoucherSegmentNotEligible    VoucherRejection = "SEGMENT_NOT_ELIGIBLE"
)

// VoucherCheck is the context a redemption is judged against.
type VoucherCheck struct {
	Voucher               Voucher
	Package               CreditPackage
	MemberIsNew           bool
	MemberRedemptionCount int
	TotalRedemptionCount  int
	Now                   time.Time
}

// ValidateVoucher returns the discount in IDR, or the reason for refusal.
// The discount can never exceed the package price.
func ValidateVoucher(c VoucherCheck) (int64, VoucherRejection) {
	v := c.Voucher
	switch {
	case v.Status != VoucherActive:
		return 0, VoucherNotActive
	case c.Now.Before(v.StartsAt):
		return 0, VoucherNotStarted
	case c.Now.After(v.EndsAt):
		return 0, VoucherEnded
	case v.UsageLimit != nil && c.TotalRedemptionCount >= *v.UsageLimit:
		return 0, VoucherUsageLimitReached
	case v.PerMemberLimit != nil && c.MemberRedemptionCount >= *v.PerMemberLimit:
		return 0, VoucherPerMemberLimitReached
	case v.ApplicablePackageIDs != nil && !contains(v.ApplicablePackageIDs, c.Package.ID):
		return 0, VoucherPackageNotEligible
	case v.EligibleSegment == SegmentNewMembers && !c.MemberIsNew:
		return 0, VoucherSegmentNotEligible
	}

	if v.Type == VoucherFixedIDR {
		return min64(v.Value, c.Package.PriceIDR), ""
	}
	discount := int64(math.Round(float64(c.Package.PriceIDR) * float64(v.Value) / 100))
	return min64(discount, c.Package.PriceIDR), ""
}

func min64(a, b int64) int64 {
	if a < b {
		return a
	}
	return b
}
